### ISSUE TITLE
Port twisted.names.dns to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6057]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6057|
|Type|enhancement|
|Created|2012-10-04 18:04:27Z|
|Branch|https://github.com/twisted/twisted/tree/dns-py3-6057|

This is needed for `twisted.names.client`.

<details><summary>Searchable metadata</summary>

```
trac-id__6057 6057
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_dns_py3_6057 branches/dns-py3-6057
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1349373867000000 1349373867000000
changetime__1349785075000000 1349785075000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35965]) Branching to 'dns-py3-6057'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @itamarst|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dns-py3-6057)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

Thanks for dealing with this.

  1. This is the 2nd time you've implemented a bytes version of `chr()`, isn't it? I suggest opening a ticket for putting this in `compat`, though actually doing the ticket isn't urgent.
  1. Returning unicode from `Charstr.__str__` etc. on Python 2 seems like a bug.
  1. I love the thing where suffix is unicode but prefix is bytes. Oh well.
  1. News file.
  1. Might be worth doing a coverage test, and maybe opening separate tickets for rarely used classes that need better test coverage before porting. I caught at least the following:
    1. Line 388 seems like `''` should be bytes. Missing test to cover this?
    1. Line 1071 (`Record_A6.decode`) looks like you're concatenating unicode and bytes. Again, missing test? Record_A6 seems undertested in general.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

> This is the 2nd time you've implemented a bytes version of chr(), isn't it? I suggest opening a ticket for putting this in compat, though actually doing the ticket isn't urgent.

Filed [#6070](https://github.com/twisted/twisted/issues/6070).

> Returning unicode from Charstr.__str__ etc. on Python 2 seems like a bug.

But it works.  So...?

> I love the thing where suffix is unicode but prefix is bytes. Oh well.

Hee hee.  But it basically makes sense, since the suffix and prefix are of difference types in the abstract, too.  Also no one uses A6 records. :/

> News file.

Misc added in r35996.

> Line 388 seems like '' should be bytes. Missing test to cover this?

Hummm.  That's actually covered by the tests.  It works because the only use of the `''` value after it is assigned to the `name` local is a test for truth to decide whether to continue the loop.  I agree it's confusing that it's a text string, though.  Perhaps it should even be `None` instead of `''` or `b''`.  I tweaked the code in r35997.

> Line 1071 (Record_A6.decode) looks like you're concatenating unicode and bytes. Again, missing test? Record_A6 seems undertested in general.

Crap.  Almost _all_ the record classes are missing test coverage for encoding and decoding.  I added some in r35998, and also fixed the `Record_A6` bug.

[Updated build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dns-py3-6057)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Thanks for the improved coverage.

  1. Removing unicode still seems wrong to me, since it violates the API and might break sometime. How about `nativeString()`? Does the right thing on both 2 and 3.
  1. Add a minimal test or two for `str2time`, which looks like it's biggest piece of missing coverage.

Once you've done that, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Removing unicode still seems wrong to me, since it violates the API and might break sometime. How about nativeString()? Does the right thing on both 2 and 3.

Removing it from what?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Sorry. Lack of sleep is getting to me.

Returning unicode from `__str__` still seems wrong to me, since it violates the API and might break something or other. How about nativeString()? Does the right thing on both 2 and 3, and seems to be exactly what you want.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [36010]) Use nativeString to return bytes on Python 2 and text on Python 3 (though Python 2 is happy with text here too, through implicit encoding I guess)

refs [#6057](https://github.com/twisted/twisted/issues/6057)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [36011]) Add some tests for str2time

refs [#6057](https://github.com/twisted/twisted/issues/6057)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Okay, did those things.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [36012]) Okay, I added this bogus documentation in this branch, guess I'll fix it here too.

refs [#6057](https://github.com/twisted/twisted/issues/6057)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [36013]) Slightly more fixing

refs [#6057](https://github.com/twisted/twisted/issues/6057)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [36014]) Merge dns-py3-6057

Author: exarkun
Reviewer: itamarst
Fixes: [#6057](https://github.com/twisted/twisted/issues/6057)

Port `twisted.names.dns` to Python 3.  As part of this, add some missing test coverage
and fix minor comparison issues.
